v2/tools/generator/internal/codegen/pipeline/add_secrets.go

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/astmodel"
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/config"
+)
+
+// AddSecretsStageID is the unique identifier for this pipeline stage
+const AddSecretsStageID = "addSecrets"
+
+// AddSecrets replaces properties flagged as secret with genruntime.SecretReference
+func AddSecrets(_ *config.Configuration) Stage { // TODO: Will need this configuration parameter eventually to allow config overrides
+	return MakeStage(
+		AddSecretsStageID,
+		"Replace properties flagged as secret with genruntime.SecretReference",
+		func(ctx context.Context, state *State) (*State, error) {
+
+			updatedSpecs, err := transformSpecSecrets(state.Types())
+			if err != nil {
+				return nil, err
+			}
+
+			updatedStatuses, err := removeStatusSecrets(state.Types())
+			if err != nil {
+				return nil, err
+			}
+
+			result := state.Types().OverlayWith(astmodel.TypesDisjointUnion(updatedSpecs, updatedStatuses))
+			return state.WithTypes(result), nil
+		})
+}
+
+func transformSpecSecrets(types astmodel.Types) (astmodel.Types, error) {
+	specVisitor := astmodel.TypeVisitorBuilder{
+		VisitObjectType: transformSecretProperties,
+	}.Build()
+
+	specTypes, err := astmodel.FindSpecConnectedTypes(types)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't find all spec types")
+	}
+
+	result := make(astmodel.Types)
+
+	for _, def := range specTypes {
+		updatedDef, err := specVisitor.VisitDefinition(def, nil)
+		if err != nil {
+			return nil, errors.Wrapf(err, "visiting type %q", def.Name())
+		}
+
+		result.Add(updatedDef)
+	}
+
+	return result, nil
+}
+
+func removeStatusSecrets(types astmodel.Types) (astmodel.Types, error) {
+	specVisitor := astmodel.TypeVisitorBuilder{
+		VisitObjectType: removeSecretProperties,
+	}.Build()
+
+	statusTypes, err := astmodel.FindStatusConnectedTypes(types)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't find all status types")
+	}
+
+	result := make(astmodel.Types)
+
+	for _, def := range statusTypes {
+		updatedDef, err := specVisitor.VisitDefinition(def, nil)
+		if err != nil {
+			return nil, errors.Wrapf(err, "visiting type %q", def.Name())
+		}
+
+		result.Add(updatedDef)
+	}
+
+	return result, nil
+}
+
+func isTypeString(t astmodel.Type) bool {
+	primType, ok := astmodel.AsPrimitiveType(t)
+	return ok && primType == astmodel.StringType
+}
```

Comment:
Add a predicate to `PropertyDefinition` called `HasType()` and use that instead of creating a standalone function.

Reply:
Did this but then realized that all that was doing was calling `prop.PropertyType().Equals`. It felt silly to have a method on property which just proxied a call to `PropertyType()` (which is already exported).

I just moved to do `prop.PropertyType().Equals` instead.

Reply:
The problem here is that we need to call `AsPrimitiveType` to perform `MetaType` unwrapping, which feels like a very powerful/risky operation. The name `HasType` doesn't really match the behavior because it sounds like it's checking if the type equals something (also accomplishable by `prop.PropertyType().Equals()`), but in reality in order to work here it would actually have to be doing something like `astmodel.Unwrap(prop.PropertyType()).Equals()`.

We previously didn't have a generic `astmodel.Unwrap` though - I've added one and done `astmodel.Unwrap(prop.PropertyType()).Equals()` directly here instead, which avoids the custom function but also doesn't promote the pattern into a method on `PropertyDefinition`.

Let me know what you think of that approach.